### DISCUSSION TITLE
About `making` Eloquent event

### DISCUSSION THREAD

**Discussion by kudashevs** [laravel/framework#53785] [Ideas]
The idea is to add a `making` event to the Eloquent model which is going to be bound to the `Model::make()` method (as the `creating` event is bound to the `Model::create()` method). Why the framework needs it?

**Use case**
Let's assume, there is a Post model and we want to automatically generate a slug from a title (_when a slug is not provided_). I tried to implement this functionality [through Mutators](https://github.com/laravel/framework/discussions/53785#discussioncomment-11650539) (but it doesn't work because you need to call a `slug` explicitly to mutate it) and through the constructor (but I didn't find a way to access the models' attributes in there - it seems there is no way to do it). 

The only working way, which I've found, is to use the Model events with closures to adjust the slug.
```php
class Post extends Model
{
    use HasFactory;

    protected $fillable = [
        'title',
        'slug',
    ];

    protected static function booted()
    {
        static::creating(function ($tag) {
            $tag->slug = str()->slug($tag->slug ?? $tag->title);
        });
    }
}
```

This works for the `create()` method, 
```
// \App\Models\Post::create(['title' => 'new title']);
App\Models\Post {
    title: "new title",
    slug: "new-title",
  }
```
but this does not work for the `make()` method.
```
// \App\Models\Post::make(['title' => 'new title']);
App\Models\Post {
    title: "new title",
  }
```
, it's not possible to have the symmetrical behavior for `Post::make()` and `Post::create()` because no events are bound to `make`.

Btw, my use case is not unique. There are other cases related to the dynamic attributes generation, for example:
- default [attributes with random values](https://stackoverflow.com/questions/54043436/default-attribute-with-random-values-in-laravel-model)
- generation of [unique codes for each user](https://blog.damirmiladinov.com/laravel/building-laravel-referral-system.html)

I would like to propose to add the `making` event and bound it to the `Model::make()` method. It seems logical for this event to exist.

**Comment by ahinkle** [comment]
What's your use case for a theoretical `making` event?

Typically `saving`, `saved`, `created`, or `creating` satisfies everything I've come across in applications.

**Reply by kudashevs** [reply]
@ahinkle Thanks for asking.

I tried to explain a use case with Post model and slug in the question. It turned out messy without a code example.
(**Update**: I just updated the question with a use case code example).

So, the other use-case. We have a model where the slug should start with a specific prefix (`/product/` in this case) no matter what. 
I see two possible ways of implementing this requirement: - via mutators, - via events. Let's compare them.

Let's implement it **via mutators**:
```php
class ProductMutators extends Model
{
    use HasFactory;

    private const PREFIX = '/product/';

    protected function slug(): Attribute
    {
        return Attribute::make(
            get: function ($value) {
                return self::PREFIX . ltrim($value, '/');
            },
            set: function ($value) {
                return self::PREFIX . ltrim($value, '/');
            }
        );
    }
}
```

Now, when the model is made, created, or updated, it has the same behavior (even in tests - this is important).
```
// \App\Models\ProductMutators::make(['title' => 'test', 'slug' => 'test']);
App\Models\ProductMutators {
    title: "test",
    slug: "/product/test",
  }
```

```
// \App\Models\ProductMutators::create(['title' => 'test', 'slug' => 'test']);
App\Models\ProductMutators {
    title: "test",
    slug: "/product/test",
    updated_at: "2024-12-10 08:28:55",
    created_at: "2024-12-10 08:28:55",
  }
```

``` 
// \App\Models\ProductMutators::factory()->make();
App\Models\ProductMutators {
    title: "Molestias ut.",
    slug: "/product/molestias-ut",
  }
```

Let's implement the same **via events**:
```php
class ProductEvents extends Model
{
    use HasFactory;

    private const PREFIX = '/product/';

    protected static function booted()
    {
        static::creating(function ($tag) {
            $tag->slug = self::PREFIX . ltrim($tag->slug ?? $tag->title, '/');
        });
        // ....
    }
}
```

``` 
// \App\Models\ProductEvents::make(['title' => 'test', 'slug' => 'test']);
App\Models\ProductEvents {
    title: "test",
    slug: "test",
}
```

``` 
// \App\Models\ProductEvents::create(['title' => 'test', 'slug' => 'test']);
App\Models\ProductEvents {
    title: "test",
    slug: "/product/test",
    updated_at: "2024-12-10 08:37:00",
    created_at: "2024-12-10 08:37:00",
  }
```

And, in tests:
```
//  \App\Models\ProductEvents::factory()->create();
App\Models\ProductEvents {
    title: "Laudantium nihil.",
    slug: "/product/laudantium-nihil",
    updated_at: "2024-12-10 08:52:01",
    created_at: "2024-12-10 08:52:01",
  }
```

```
// \App\Models\ProductEvents::factory()->make();
App\Models\ProductEvents {
    title: "Ex fugit qui.",
    slug: "ex-fugit-qui",
  }
```

As you can see, it changes the `slug` attribute on `create()`. But, there is no way to change the `slug` attribute on `make()`.
Not being able to do this in tests is crucial for me. And, it feels likes the `make()` is a little bit deprived 😄 

**Update**: this use case might seem like overengineering, however, I wanted to emphasize the difference between Mutators and Events. When we are talking about a use case for a `making` event, we are talking about the situations were we need to access the other fields/properties of the model during the creation process. In terms of the above-mentioned example, imagine we have a model where the slug should start with a specific prefix `/product-year/` and the `year` in this prefix is based on `created_at` property.

**Reply by henzeb** [reply]
If you can solve it with attributes/mutators, the chances of this making events getting in are slim...

Also, events have a different use case then modifying its own record data. It's more for sending emails, clearing/updating caches, that kind of stuff.

Unless you need the exact timestamp including the time and seconds, mutators icm Carbon will do the trick

**Reply by kudashevs** [reply]
> If you can solve it with attributes/mutators, the chances of this making events getting in are slim...

I haven't said that. In the main question I said the opposite: ```I tried to implement this functionality through Mutators (but it doesn't work because you need to call a slug explicitly to mutate it) and through the constructor (but I didn't find a way to access the models' attributes there).```

> Also, events have a different use case then modifying its own record data. It's more for sending emails, clearing/updating caches, that kind of stuff.
> 
> Unless you need the exact timestamp including the time and seconds, mutators icm Carbon will do the trick

Could you please show, how to implement this slug auto-generation functionality with Mutators and Carbon and without reloading the Model after each update? Probably, I've missed something.

This is the use case with a slug generation:
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Product extends Model
{
    protected $fillable = [
        'title',
        'slug',
        'published_at',
    ];

    protected static function booted()
    {
        static::creating(function ($tag) {
            $slug = str()->slug($tag->slug ?? $tag->title);
            $year = (isset($tag->published_at)) ?  $tag->published_at->format('Y') : now()->format('Y');
            $tag->slug = '/products/product-line-' . $year . '-' . ltrim($slug, '/');
        });
        static::updating(function ($tag) {
            $slug = str()->slug($tag->title);
            $year = $tag->published_at->format('Y');
            $tag->slug = '/products/product-line-' . $year . '-' . ltrim($slug, '/');
        });
    }
}
```

Product creation:
```php
$product = \App\Models\Product::create(['title' => 'Unique product', 'published_at' => now()]);
```
results into:
```php
= App\Models\Product {
    title: "Unique product",
    published_at: Illuminate\Support\Carbon @1734957945 {#5036
      date: 2024-12-23 12:45:45.110753 UTC (+00:00),
    },
    slug: "/products/product-line-2024-unique-product",
  }
```

Product update:
```php
$product->title = 'New product';
$product->save();
 ```
results into:
```php
 = App\Models\ProductEvents {
    title: "New product",
    published_at: Illuminate\Support\Carbon @1734957945 {#5036
      date: 2024-12-23 12:45:45.110753 UTC (+00:00),
    },
    slug: "/products/product-line-2024-new-product",
  }
```

 I would also appreciate, if you could show how to do it with the `Product::make()` method.

**Reply by henzeb** [reply]
The mutator should be on the "name" attribute. When the name changes, the slug will update too.

Please be aware that we shouldn't develop for tests, but develop so it is testable ...

**Reply by kudashevs** [reply]
> The mutator should be on the "name" attribute. When the name changes, the slug will update too.

Thank you for explaining so thoroughly. However, as I've said, it didn't work in my case. So, I would like to see the code, if you don't mind - not the explanations.

**Reply by kudashevs** [reply]
> Please be aware that we shouldn't develop for tests, but develop so it is testable ...

Thank you for pointing it out. However, we are not talking about testing. We're talking about the Model's behavior.

If the `create` and `make` methods are responsible for creating a Model's instance, they should treat the information about the instance in the same way and, ideally, provide the same functionality to change the provided data. Right now, there is the discrepancy.

**Reply by henzeb** [reply]
So we don't want extra events to listen to, we want create and make also use the mutators where needed? You spotted a possible bug?

**Reply by ahinkle** [reply]
https://laravel.com/docs/11.x/eloquent-mutators#defining-a-mutator

```php
public function setSlugAttribute($value)
{
    $this->attributes['slug'] = $value ?? str()->slug($this->title);
}
```

**Reply by kudashevs** [reply]
> https://laravel.com/docs/11.x/eloquent-mutators#defining-a-mutator
> 
> ```
> public function setSlugAttribute($value)
> {
>     $this->attributes['slug'] = $value ?? str()->slug($this->title);
> }
> ```

@ahinkle Have you tried your code example? It simply doesn't work with the `create` only.

Let's try both ways of writing Mutators (the old one and the new one). It will be clearer what I'm talking about .

### The old one
```php
    public function setSlugAttribute($value)
    {
        var_dump('triggered');
        $slug = $value ?? str()->slug($this->title);
        $this->attributes['slug'] = $slug;
    }
```

Product creation:
```php
$product = \App\Models\Product::create(['title' => 'Unique product']);
```
results into:
```php
Illuminate\Database\QueryException  SQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed
: products.slug (Connection: sqlite, SQL: insert into "products" ("title", "published_at", "updated_at", "created_at"
) values (Unique product, 2024-12-23 13:59:58, 2024-12-23 13:59:58, 2024-12-23 13:59:58)).
```

### The new one
```php
    protected function slug(): Attribute
    {
        return Attribute::make(
            set: function ($value, array $attributes) {
                var_dump('triggered');
                $slug = $value ?? str()->slug($attributes['title']);
                $this->attributes['slug'] = $slug;
            },
        );
    }
```
results into:
```php
  Illuminate\Database\QueryException  SQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed
: products.slug (Connection: sqlite, SQL: insert into "products" ("title", "published_at", "updated_at", "created_at"
) values (Unique product, 2024-12-23 14:02:13, 2024-12-23 14:02:13, 2024-12-23 14:02:13)).
```

So, as you can see, **in both cases the Mutators weren't even triggered**.
To make your code work (as I described earlier) you must access the property directly.

```php
> $product = \App\Models\Product::make(['title' => 'Unique product']);
= App\Models\Product {
    title: "Unique product",
  }
```
then it will be mutated.
```php
 $product->slug;
 // string(9) "triggered"
App\Models\Product {
    title: "Unique product",
    slug: "unique-product",
  }
 ```
 
 If you don't access it, it doesn't mutate. So, the only possible way to make it work, which I've found, is through events.
 
 Did I miss something?

**Reply by kudashevs** [reply]
> So we don't want extra events to listen to, we want create and make also use the mutators where needed? You spotted a possible bug?

Sorry, I don't quite understand your question. Who are `we`? Why we don't want extra events? 
I wrote the opposite, I want this extra event. The event looks promising and it will be useful for the community.

**Reply by henzeb** [reply]
I am trying to understand what you are trying to do. You want to have a title that has a slug created on "make". That doesn't work the way you thought it would work and now you want events. you don't need events.

```php
class Post extends Model  {

    protected $fillable = ['name'];

    public function name(): Attribute
    {
        return Attribute::make(
            set: fn(string $name) => [
                'name' => $name,
                'slug' => Str::slug($name)
            ]
        );
    }
}

dd(
    Post::make([
        'name' => 'my name',
    ])->toArray()
);

# not touching it: 
#  array:2 [
#   "name" => "my name"
#   "slug" => "my-name"
# ]

dd(
    Post::make([
        'name' => 'my name',
    ])->slug
);

# output: "my-name"

```

I am pretty sure this will do what you wanted it to do. I am also pretty confident by setting the attribute in the "get" when empty/not set also works for the other use cases with random values and unique values .

**Reply by kudashevs** [reply]
> I am trying to understand what you are trying to do. You want to have a title that has a slug created on "make". That doesn't work the way you thought it would work and now you want events. you don't need events.

@henzeb Answering your question. I try to make a `slug` being generated from the provided `title` whenever the Model is made (e.g. `Model::make()`).

I really want to appreciate your help and your solution that kind of work. But it works! Thank you very much for providing it!

Why the solution kind of works:
- the first reason: it is based on mutating the `name` (`title` or whatever), but not the `slug`. One may say that this is a nitpicking, but to me it looks kind of odd (we mutate the `name` to have the `slug` matated - not really obvious)
- the second reason: ~~this solution does not work in Tinker (I tried both 10 and 11 versions)~~. It does work in Tinker.

Other than that, this is a working solution and I'm going to use it as a workaround. Thank you very much, once again!

**Comment by kudashevs** [comment]
Since there is no much interest in the new event and there is a workaround (thanks to @henzeb) , I better close the discussion.
